---
title: "Gene Expression in Cancer Cells"
author: "Daniel Kaplan"
date: "CVC Workshop, July 2014"
output: pdf_document
---

```{r include=FALSE}
require(DCFdevel)
require(knitr)
require(mosaic)
ShowAnswers=TRUE
opts_chunk$set( tidy=FALSE, include=ShowAnswers, prompt=FALSE, comment=NA)
options(width=120)
```

## Gene Expression and Gene Expression Studies

For those unfamiliar with gene expression, we include the following description:

> Gene expression is the process by which information from a gene is used in the synthesis of a functional gene product. These products are often proteins, but in non-protein coding genes such as ribosomal RNA (rRNA), transfer RNA (tRNA) or small nuclear RNA (snRNA) genes, the product is a functional RNA... Several steps in the gene expression process may be modulated, including the transcription, RNA splicing, translation, and post-translational modification of a protein. Gene regulation gives the cell control over structure and function, and is the basis for cellular differentiation, morphogenesis and the versatility and adaptability of any organism... The timing, location, and amount of gene expression can have a profound effect on the functions (actions) of the gene in a cell or in a multicellular organism.
In genetics, gene expression is the most fundamental level at which the genotype gives rise to the phenotype. The genetic code stored in DNA is "interpreted" by gene expression, and the properties of the expression give rise to the organism's phenotype. Such phenotypes are often expressed by the synthesis of proteins that control the organism's shape, or that act as enzymes catalysing specific metabolic pathways characterising the organism.

> \hfill Source: [http://en.wikipedia.org/wiki/Gene_expression](http://en.wikipedia.org/wiki/Gene_expression)

> Developed in the 1990s, DNA microarrays have revolutionized the way in which gene expression is now analyzed by allowing the RNA products of thousands of genes to be monitored at once. By examining the expression of so many genes simultaneously, we can now begin to identify and study the gene expression patterns that underlie cellular physiology: we can see which genes are switched on (or off) as cells grow, divide, or respond to hormones or to toxins.

> DNA microarrays are little more than glass microscope slides studded with a large number of DNA fragments, each containing a nucleotide sequence that serves as a probe for a specific gene. The most dense arrays may contain tens of thousands of these fragments in an area smaller than a postage stamp, allowing thousands of hybridization reactions to be performed in parallel (Figure 8-62). Some microarrays are generated from large DNA fragments that have been generated by PCR and then spotted onto the slides by a robot. Others contain short oligonucleotides that are synthesized on the surface of the glass wafer with techniques similar to those that are used to etch circuits onto computer chips. In either case, the exact sequence—and position—of every probe on the chip is known. Thus any nucleotide fragment that hybridizes to a probe on the array can be identified as the product of a specific gene simply by detecting the position to which it is bound.

> \hfill Source: [www.ncbi.nlm.nih.gov/books/NBK26818/](http://www.ncbi.nlm.nih.gov/books/NBK26818/)

## NCI 60

In the 1980s, the [National Cancer Institute](http://www.cancer.gov/) developed a set of 60 cancer cell lines, called [NCI60](http://www.broadinstitute.org/mpr/NCI60/NCI60.html). The original purpose was for screening potential anti-cancer drugs. A recent overview in [*Nature*](http://www.nature.com/nrc/journal/v6/n10/full/nrc1951.html) describes some of the ongoing work with these cell lines.  

Here you will examine gene expression in these cell lines using data described in [Staunton *et al.*](http://www.pnas.org/content/98/19/10787.full). More than 41,000 probes were used for each of the 60 cell lines.

For convenience, the data are provided by the `DCIdevel` package.
```{r}
# load required packages
require(DCFdevel)
require(mosaic)
# this will load two data frames: nci60expression and nci60celline
data(nci60)       
```

The `nci60` data tables are lightly re-organized from the [form provided by Staunton *et al.*](http://www.broadinstitute.org/mpr/NCI60/NCI60.html).

Two separate data frames are provided: `nci60expression` and `nci60cellLine`.  The expression data is quite big --- 41,078 probes by 60 cell lines --- so there's little point in looking at it directly:
```{r}
dim(nci60expression)
```

Each of the 2,454,680 numbers in the expression data is on a log scale.


To help visualize how the process of data processing, here is a small subset of 15 cases and 6 cell lines from the data.  These are presented in the same format as the original.  The name of the cell line starts with two or three letters indicating the tissue type: `BR` is breast, `CNS` is brain (central nervous system), `CO` is colon.
```{r include=TRUE}
set.seed(300)
row.names(nci60expression) <- make.names(nci60expression$Probe, unique=TRUE)
Small <- nci60expression %>%
  select(c(1,2,4,7,8,14,15)) %>%       # only these columns
  sample(size=15, orig.ids=FALSE)    # 15 random rows
```
```{r}
print(Small, row.names=FALSE)
```

Construct this small dataset for developing your computer statements.

Once you have the `small` dataframe, it's relatively easy to display how the data indicate the cell lines are related.
```{r}
Small2 <- Small %>% select(-Probe)
SmallDend <- Small2 %>% dist() %>% hclust()
mplot(SmallDend, data=Small2, k=3, heatmap=0.5, labels=TRUE)
```

But this method wouldn't scale nicely to all 41,000 probes.  You'll have to do some work to get the data in form where you can construct the dendrograms.


## Narrow vs. Wide Data Formats

* The expression data is provided in *wide* format.  What would it look like in *narrow* format.

* Create the *narrow* format data, using `Expression` as the name of that variable in the result and `CellLine` as the variable name for that information.
    `melt()` let's you specify the `value.name` and the `variable.name` as character strings.  You'll have to figure out which is which.  
```{r}
Narrow <- melt(Small, value.name='Expression', variable.name='CellLine')
```

* If the expression does not vary much for each cell line across probes, that cell line might be faulty.  One way to measure the variation is with the [standard deviation](http://en.wikipedia.org/wiki/Standard_deviation) (`sd()`) Make a chart to compare the variation in expression across probes in each of the cell lines.
    Explain why or why not your chart form would be effective for the full data.

```{r}
NarrowByCellLine <- 
Narrow %>% 
  group_by( CellLine ) %>%
  summarise( spread=sd(Expression), n=n() ) 

ggplot(data=NarrowByCellLine, aes(x=CellLine, y=spread)) + 
  geom_bar(fill='grey', stat='identity')
```



    
## Weeding Out

The variation of expression of a probe across cell lines can be used as an indicator of the relationship between the expression of that gene and cancer types.  But there are many probes that hardly vary across the cell lines.  For statistical reasons, to increase the "power" of a study, it's sensible to delete these probes from the data.  The expression shown by most of the probes do not vary across cell lines.


* Find the standard deviation of the expression for each probe across cell lines. 
    * Decide whether you want to do this from the data in long or wide format.  Either is possible.
```{r}
Narrow %>% 
  group_by( Probe ) %>%
  summarize( spread=sd(Expression)) -> NarrowByProbe
```
    
* Make a bar chart of `small` that shows how the expression varies across cell lines for each probe. 
    * Explain why or why not your chart form would be effective for the full data.
    * What do you think would be the best ordering of probes for the purpose of spotting how many probes have high variation?
    
    ```{r}
    ggplot(data=NarrowByProbe, aes(x=reorder(Probe, spread), y=spread)) + 
      geom_bar(fill='gray', stat='identity')
    ```

* Make a histogram and/or density plot showing how probes differ in terms of their variation across cell lines.
    *Explain why or why not this cart would be effective for the full data.
    
    ```{r}
    ggplot(data=NarrowByProbe, aes(x=spread)) + 
    geom_histogram(fill='gray', binwidth=0.5, aes(y= ..density..)) + 
    geom_density(aes(y=..density..))
    ```
* Make a dataset containing the expression data for the five probes with the highest standard deviation.
    ```{r}
    Highest <- 
      NarrowByProbe %>%
      filter(rank(desc(spread)) <= 5)
    ```

* Join the highest varying probes with the `small` dataset.
    * Experiment with the different sorts of join to find one that keeps just the highest varying probes in `small`.
    * When you have a working result, use `select()` to get rid of the `spread` variable.  It's done its work and is no longer needed.
    ```{r}
    Keepers <- Small %>% inner_join(Highest) %>% select(-spread)
    ```

The dendrogram/heat-map will look like this:
```{r}
KeepersTmp <- Keepers %>% select(-Probe) 
KeepersDend <- KeepersTmp %>% dist() %>% hclust()
mplot(KeepersDend, data=KeepersTmp, k=3, heatmap=0.5, labels=TRUE)
```

You might prefer to create a dendrogram of the cell types rather than the probes.  This is a matter of taking the transpose of the data.  Here's how:
```{r echo=ShowAnswers, include=TRUE}
KeepersTranspose <- Keepers %>% select(-Probe) %>% t(.)
colnames(KeepersTranspose) <- Keepers$Probe
KeepersTransposed <- as.data.frame(KeepersTranspose)
```

And the corresponding dendrogram:

```{r}
Dend <- hclust(dist(KeepersTransposed))
mplot(Dend, data=KeepersTransposed, k=3, heatmap=0.5, labels=TRUE)
```

This dendrogram doesn't show a good classification of the tissue types, but there isn't much data going into it.  More data might produce a better classification.


## On Your Own

When you have the above working ...

For the whole data set --- picking out just 100 or so probes with the greatest variation ---
make dendrograms of 

* the cell lines using probe expression
* the probes using cell lines

Pick out one of the top varying probes.

* Merge the `tissue` variable based on `cellLine`.  Construct a "dynamite plot" showing how the probe expression varies for the different tissues.

